vyper/venom/passes/remove_unused_variables.py

```diff
@@ -1,47 +1,42 @@
+from vyper.utils import OrderedSet
 from vyper.venom.analysis.dfg import DFGAnalysis
-from vyper.venom.analysis.liveness import LivenessAnalysis
-from vyper.venom.basicblock import IRBasicBlock
+from vyper.venom.basicblock import IRInstruction
 from vyper.venom.passes.base_pass import IRPass
 
 
 class RemoveUnusedVariablesPass(IRPass):
-    def run_pass(self):
-        self.analyses_cache.request_analysis(LivenessAnalysis)
-
-        for bb in self.function.get_basic_blocks():
-            self._remove_unused_variables(bb)
+    """
+    This pass removes instructions that produce output that is never used.
+    """
 
-        self.analyses_cache.invalidate_analysis(DFGAnalysis)
+    dfg: DFGAnalysis
+    work_list: OrderedSet[IRInstruction]
 
-    def _remove_unused_variables(self, bb: IRBasicBlock):
-        """
-        Remove the instructions of a basicblock that produce output that is never used.
-        """
-        i = 0
-        while i < len(bb.instructions) - 1:
-            inst = bb.instructions[i]
-            i += 1
+    def run_pass(self):
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
 
-            # Skip volatile instructions
-            if inst.volatile:
-                continue
+        work_list = OrderedSet()
+        self.work_list = work_list
 
-            # Skip instructions without output
-            if inst.output is None:
-                continue
+        for _, inst in self.dfg.outputs.items():
+            work_list.add(inst)
 
-            # Skip instructions that produce output that is used
-            next_liveness = bb.instructions[i].liveness
-            if inst.output in next_liveness:
-                continue
+        while len(work_list) > 0:
+            inst = work_list.pop()
+            self._process_instruction(inst)
 
-            # Remove the rest
-            del bb.instructions[i - 1]
+    def _process_instruction(self, inst):
+        if inst.output is None:
+            return
+        if inst.volatile:
+            return
+        uses = self.dfg.get_uses(inst.output)
+        if len(uses) > 0:
+            return
 
-            # backtrack to the *previous* instruction, in case we removed
-            # an instruction which had prevented the previous instruction
-            # from being removed
-            i -= 2
+        for operand in inst.get_inputs():
+            new_uses = self.dfg.remove_use(operand, inst)
+            for use in new_uses:
```

Comment:
can use `self.work_list.addmany(new_uses)`

Reply:
Even I can’t keep up with all the cool utility functions I write 😅